google/cloud/spark_connect/session.py

```diff
@@ -474,6 +476,70 @@ def _remove_stoped_session_from_file(self):
                     f"Exception while removing active session in file {file_path} , {e}"
                 )
 
+    def addArtifacts(
+        self,
+        *artifact: str,
+        pyfile: bool = False,
+        archive: bool = False,
+        file: bool = False,
+        pypi: bool = False,
+    ) -> None:
+        """
+        Add artifact(s) to the client session. Currently only local files & pypi installations are supported.
+
+        .. versionadded:: 3.5.0
+
+        Parameters
+        ----------
+        *path : tuple of str
+            Artifact's URIs to add.
+        pyfile : bool
+            Whether to add them as Python dependencies such as .py, .egg, .zip or .jar files.
+            The pyfiles are directly inserted into the path when executing Python functions
+            in executors.
+        archive : bool
+            Whether to add them as archives such as .zip, .jar, .tar.gz, .tgz, or .tar files.
+            The archives are unpacked on the executor side automatically.
+        file : bool
+            Add a file to be downloaded with this Spark job on every node.
+            The ``path`` passed can only be a local file for now.
+        pypi : bool
+            This option is only available with GoogleSparkSession. eg. `spark.addArtifacts("spacy==3.8.4", "torch",  pypi=True)`
+            Installs PyPi package (with its dependencies) in the active Spark session on the driver and executors.
+
+        Notes
+        -----
+        This is an API dedicated to Spark Connect client only. With regular Spark Session, it throws
+        an exception.
+        Regarding pypi: Popular packages are already pre-installed in s8s runtime.
+        https://cloud.google.com/dataproc-serverless/docs/concepts/versions/spark-runtime-2.2#python_libraries
+        If there are conflicts/package doesn't exist, it throws an exception.
+        """
+        if sum([pypi, file, pyfile, archive]) > 1:
+            raise ValueError(
+                "'pyfile', 'archive', 'file' and/or 'pypi' cannot be True together."
+            )
+        if pypi:
+            to_install = set([p for p in artifact]).difference(
+                self._installed_artifacts
+            )
+            if len(to_install) == 0:
+                logger.info(
+                    "Ignoring as all artifacts have already been added earlier"
```

Comment:
pip will ignore them by itself, so likely we do not need to handle this on the client?

Reply:
Yes pip will ignore. However, in a Notebook, cells are regularly run multiple times. I am trying to avoid making multiple calls (each with different file) and also waiting for pip install process call on server side adding extra delay. There can be transient network issues as well (the more calls we make) etc.

In terms of implementation, we are not doing much complex. We are just maintaining a set() of string, taking a diff and clearing it up when session is stopped. I have added UTs to cover all scenarios.

Therefore, I am in favour of adding this.

Reply:
I think that this caching may have unintended behavior that maybe confusing. For example, if customer wants to run install for the same package to update it to the latest version then it will not get update due to the cache? Overall it looks like pre-mature optimization that may have unintended consequence that we may know about. Also, I think that Jupyter Sessions do not have this functionality too, so it may be better to keep them consistent.